src/ocrd/processor/base.py

```diff
@@ -61,17 +68,49 @@ class Processor():
     a number of optional or mandatory parameters.
     """
 
+    @property
+    def metadata(self):
+        """the ocrd-tool.json dict of the package"""
+        if hasattr(self, '_metadata'):
+            return self._metadata
+        self._metadata = json.loads(resource_string(self.__module__.split('.')[0], 'ocrd-tool.json'))
+        return self._metadata
+
+    @property
+    def version(self):
+        """the version of the package"""
+        if hasattr(self, '_version'):
+            return self._version
+        self._version = self.metadata['version']
+        return self._version
+
+    @property
+    def executable(self):
+        """the executable name of this processor tool"""
+        if hasattr(self, '_executable'):
+            return self._executable
+        self._executable = os.path.basename(inspect.stack()[-1].filename)
+        return self._executable
+
+    @property
+    def ocrd_tool(self):
+        """the ocrd-tool.json dict of this processor tool"""
+        if hasattr(self, '_ocrd_tool'):
+            return self._ocrd_tool
+        self._ocrd_tool = self.metadata['tools'][self.executable]
+        return self._ocrd_tool
+
     def __init__(
             self,
-            workspace : Workspace,
+            # FIXME: deprecate in favor of process_workspace(workspace)
+            workspace : Optional[Workspace],
             ocrd_tool=None,
             parameter=None,
-            # TODO OCR-D/core#274
-            # input_file_grp=None,
-            # output_file_grp=None,
-            input_file_grp="INPUT",
-            output_file_grp="OUTPUT",
+            input_file_grp=None,
+            output_file_grp=None,
             page_id=None,
+            download_files=True,
+            # FIXME: deprecate all the following in favor of respective methods
```

Comment:
Well, like I said above:
> * I did not sufficiently separate processing from non-processing context in the constructor, yet. The goal here was to first find a viable path in that direction.
>   Thus, I did add a `setup` method already, but the constructor still determines non-processing or processing context. I believe we should just use methods for that e.g. `show_resources` instead of taking that as a kwarg in the constructor. But it's an even larger change. (Hence the `fixme` comments about deprecating this.)

We could also do that larger change now, if you are so inclined.

Reply:
At this point, there is a large intersection of those who change the API and those who have to implement it for the processors. So I would actually do the (flood of) deprecation warnings now, so we can see where we need to adapt implementations. But if you want to keep the changeset more focussed, we can do that once the API changes are stable.

Reply:
If we want to get the maximum benefit of deprecation warnings and regression tests, IMO we should pre-release early – more or less based on what we have now, with ocrd_tesserocr and ocrd_kraken as they are now, but with an explicit dependency on the pre-release, and then do follow-up pre-releases as we see fit. We could even keep the PRs on both sides open and release manually. 

Changing the constructor would not necessarily mean further adaptations to the external processors – most of the work would be in ocrd_cli_wrap_processor. So I would not expect to see any actual deprecation warnings for that.

Reply:
> more or less based on what we have now, with ocrd_tesserocr and ocrd_kraken as they are now, but with an explicit dependency on the pre-release, and then do follow-up pre-releases as we see fit. We could even keep the PRs on both sides open and release manually.

Agreed, we will release `v3.0.0a1` with whatever the state is by Thursday, then keep refining the open points for a `v3.0.0b1`/`v3.0.0rc1`. With pip's versioning semantics, it won't be automatically updated unless explicitly provided in the `requirements.txt` (`>= 3` would still lead to a "Could not find a version that satisfies the requirement", `>= 3.0.0a1` will work IIUC).

We just have to be careful about the Docker images which don't have the alpha/beta/RC mechanism. I see no real harm in updating those, but in the interest of stability, we should wait with those until stable?

Reply:
> We just have to be careful about the Docker images which don't have the alpha/beta/RC mechanism. I see no real harm in updating those, but in the interest of stability, we should wait with those until stable?

CD will trigger with the merge, but no earlier (unless manually running the workflow on this branch).